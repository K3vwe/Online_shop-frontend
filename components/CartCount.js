import React from "react";
import PropTypes from "prop-types";
import { TransitionGroup, CSSTransition } from "react-transition-group";
import Styled from "styled-components";

const AnimationStyle = Styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
  }
  .count-enter {
    transform: rotateX(0);
  }
  .count-enter-active{
    transform: rotateX(0.5turn);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: rotateX(0.5turn);
  }
`;

const Dot = Styled.div`
  color: white;
  background: ${(props) => props.theme.primary};
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => (
  <AnimationStyle>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        classNames="count"
        className="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyle>
);

export default CartCount;
