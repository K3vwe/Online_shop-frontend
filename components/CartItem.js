import React from "react";
import FormatMoney from "../lib/formatMoney";
import styled from "styled-components";
import RemoveFromCart from "./RemoveFromCart";

const StyledCartItem = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => {
  if (!cartItem.item)
    return (
      <StyledCartItem>
        <p>This item has been removed</p>
        <RemoveFromCart id={cartItem.id} />
      </StyledCartItem>
    );
  return (
    <StyledCartItem>
      <img width="100px" src={cartItem.item.image} alt={cartItem.item.title} />
      <div className="cart-items-details">
        <h3>{cartItem.item.title}</h3>
        <p>
          {FormatMoney(cartItem.item.price * cartItem.quantity)}
          {" - "}
          <em>
            {cartItem.quantity}
            &times;
            {FormatMoney(cartItem.item.price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </StyledCartItem>
  );
};

export default CartItem;
