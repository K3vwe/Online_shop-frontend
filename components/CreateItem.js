import React, { useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import Router from "next/router";
import ErrorMessage from "./ErrorMessage";
import Form from "./styles/Form";
import formatMoney from "../lib/formatMoney";
import Item from "./styles/ItemStyles";

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $price: Int!
    $description: String!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

const CreateItem = () => {
  const [state, setState] = useState({
    title: "",
    description: "",
    price: 0,
  });
  // const [description, setDescription] = useState(
  //   "Its a customized ps4 pad, play with style"
  // );
  const [image, setImage] = useState("");
  const [largeImage, setLargeImage] = useState("");
  // const [price, setPrice] = useState(1000);
  const [createItem, { loading, error }] = useMutation(CREATE_ITEM_MUTATION);

  const handleChange = (event) => {
    const { name, type, value } = event.target;
    const val = type === "number" ? parseInt(value || 0, 10) : value;
    setState((state) => ({ ...state, [name]: val }));
  };

  const uploadFile = async (event) => {
    const files = event.target.files;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "sickfits");

    const res = await fetch(
      "https://api.cloudinary.com/v1_1/swissdevchi/image/upload",
      {
        method: "POST",
        body: data,
      }
    );
    const file = await res.json();

    setImage(file.secure_url);
    setLargeImage(file.eager[0].secure_url);
  };

  const { title, price, description } = state;
  return (
    <div>
      {
        <Form
          onSubmit={async (event) => {
            event.preventDefault();
            // Stop form from submitting
            const res = await createItem({
              variables: { title, description, image, largeImage, price },
            });
            // Call the Mutation
            // Change them to the single item page
            Router.push({
              pathname: "/item",
              query: { id: res.data.createItem.id },
            });
          }}
        >
          <ErrorMessage error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
            <label htmlFor="file">
              Image
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an Image"
                required
                onChange={uploadFile}
              />
              {image && <img width="200" src={image} alt={"Image preview"} />}
            </label>
            <label htmlFor="title">
              Title
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Title"
                required
                value={title}
                onChange={handleChange}
              />
            </label>
            <label htmlFor="price">
              Price
              <input
                type="number"
                id="price"
                name="price"
                placeholder="Price"
                required
                value={price}
                onChange={handleChange}
              />
            </label>
            <label htmlFor="description">
              Description
              <textarea
                id="description"
                name="description"
                placeholder="Description"
                required
                value={description}
                onChange={handleChange}
              />
            </label>
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      }
    </div>
  );
};

export default CreateItem;
