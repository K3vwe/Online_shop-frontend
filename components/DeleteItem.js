import React from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { ALL_ITEMS_QUERY } from "./Items";
import Item from "./Item";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

const DeleteItem = ({ children, id }) => {
  const [deleteItem] = useMutation(DELETE_ITEM_MUTATION, {
    variables: { id },
    update(cache, { data: { deleteItem } }) {
      // Manually update the cache on the client so it matches the server
      // 1. Read the cache for the item we want
      const { items } = cache.readQuery({ query: ALL_ITEMS_QUERY });
      // update the cache with the filtered items
      cache.writeQuery({
        query: ALL_ITEMS_QUERY,
        data: {
          items: items.filter((item) => item.id !== deleteItem.id),
        },
      });
    },
  });
  return (
    <div>
      {
        <button
          onClick={() => {
            if (confirm("Are you sure you want to delete this item")) {
              deleteItem().catch((error) => {
                alert(error.message);
              });
            }
          }}
        >
          {children}
        </button>
      }
    </div>
  );
};

export default DeleteItem;
