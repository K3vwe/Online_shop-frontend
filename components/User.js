import React from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import PropTypes from "prop-types";

export const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permission
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
          description
        }
      }
    }
  }
`;

const User = (props) => {
  const { data } = useQuery(CURRENT_USER_QUERY);
  return <div {...props}>{props.children(data)}</div>;
};

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
