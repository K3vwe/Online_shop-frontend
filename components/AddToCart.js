import React from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./User";

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ id }) => {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  return (
    <button disabled={loading} onClick={addToCart}>
      Add{loading && "ing"} to Cart ðŸ›’
    </button>
  );
};

export default AddToCart;
