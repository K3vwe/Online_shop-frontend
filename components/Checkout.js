import React from "react";
import { PaystackConsumer } from "react-paystack";
import { useMutation } from "@apollo/react-hooks";
import Router from "next/router";
import NProgress from "nprogress";
import gql from "graphql-tag";
import calcTotalPrice from "../lib/calcTotalPrice";
import SickButton from "./styles/SickButton";
import User, { CURRENT_USER_QUERY } from "./User";

export const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

const totalItems = (cart) => {
  return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0);
};

const Checkout = ({ children }) => {
  const [createOrder] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  const onToken = async (res) => {
    NProgress.start();
    const order = await createOrder({
      variables: { token: res.reference },
    }).catch((error) => {
      console.log(error.message);
    });
    Router.push({
      pathname: "/order",
      query: { id: order.data.createOrder.id },
    });
  };

  return (
    <User>
      {({ me }) => {
        const componentProps = {
          reference: new Date().getTime(),
          email: me.email,
          amount: calcTotalPrice(me.cart) * 100,
          label: "Online-Shop",
          publicKey: "pk_test_e822234b0318189bdf4d1152324516f7f5b43327",
          metadata: {
            custom_field: [
              {
                display_name: "Description",
                variable_name: "description",
                value: `Order of ${totalItems(me.cart)} items!`,
              },
            ],
          },
          onSuccess: (res) => {
            console.log("====>", onToken(res));
          },
        };

        return (
          <PaystackConsumer {...componentProps}>
            {({ initializePayment }) => {
              return (
                <SickButton
                  onClick={() => {
                    initializePayment();
                  }}
                >
                  Checkout
                </SickButton>
              );
            }}
          </PaystackConsumer>
        );
      }}
    </User>
  );
};

export default Checkout;
