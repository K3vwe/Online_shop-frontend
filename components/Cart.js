import React from "react";
import gql from "graphql-tag";
import { useMutation, useQuery } from "@apollo/react-hooks";
import CartStyles from "./styles/CartStyles";
import User from "./User";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
import Checkout from "./Checkout";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

const Cart = () => {
  const { data } = useQuery(LOCAL_STATE_QUERY);
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);

  return (
    <User>
      {(userData) => {
        if (!userData || !userData.me) return null;
        const { cart } = userData.me;
        return (
          <CartStyles open={data.cartOpen}>
            <header>
              <CloseButton onClick={toggleCart} title="close">
                &times;
              </CloseButton>
              <Supreme>Your Cart</Supreme>
              <p>
                You have {cart.length} item
                {cart.length <= 1 ? "" : "s"} in your cart.
              </p>
            </header>
            <ul>
              {cart.map((cartItem) => (
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(cart))}</p>
              <Checkout />
            </footer>
          </CartStyles>
        );
      }}
    </User>
  );
};

export default Cart;
