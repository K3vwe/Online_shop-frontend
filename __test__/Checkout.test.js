import { mount } from "enzyme";
import wait from "waait";
import Checkout from "../components/Checkout";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import { MockedProvider } from "@apollo/react-testing";
import toJson from "enzyme-to-json";

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: { ...fakeUser(), cart: [fakeCartItem()] } } },
  },
];

describe("<Checkout/>", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Checkout />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const checkoutButton = wrapper.find("ReactStripeCheckout");
    expect(toJson(checkoutButton)).toMatchSnapshot();
  });
});
