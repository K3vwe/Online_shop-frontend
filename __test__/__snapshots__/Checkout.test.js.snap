// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkout/> renders and matches snapshot 1`] = `null`;
